### PR TITLE
[AzureMonitorLiveMetrics] Add known fields to MonitoringDataPoint

### PR DESCRIPTION
**Pull request by rajkumar-rangaraj**
Following are copied from AI SDK
* `GuidExtensions.cs` 
* `IsWebAppRunningInAzure` and `GetStreamId` methods
* `AccessingEnvironmentVariableFailedWarning` event

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/SdkVersionUtils.cs [hunk 1]
```diff
@@ -53,15 +58,23 @@ internal static bool IsDistro
 
                 if (shortVersion.Length > 20)
                 {
+#if AZURE_MONITOR_EXPORTER
                     AzureMonitorExporterEventSource.Log.VersionStringUnexpectedLength(type.Name, versionString);
+#elif LIVE_METRICS_EXPORTER
+                LiveMetricsExporterEventSource.Log.VersionStringUnexpectedLength(type.Name, versionString);
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/SdkVersionUtils.cs:64]: nit; indention

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Azure.Monitor.OpenTelemetry.LiveMetrics.csproj [hunk 2]
```diff
@@ -20,11 +20,14 @@
...
+    <Compile Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Internals\AksResourceProcessor.cs" Link="Internals" />
     <Compile Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Internals\NullableAttributes.cs" LinkBase="Internals" />
     <Compile Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Internals\ExceptionExtensions.cs" LinkBase="Internals" />
-    <Compile Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Internals\SchemaConstants.cs" Link="SchemaConstants.cs" />
-    <Compile Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Internals\SemanticConventions.cs" Link="SemanticConventions.cs" />
+    <Compile Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Internals\SchemaConstants.cs" Link="Internals" />
+    <Compile Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Internals\SdkVersionUtils.cs" Link="Internals" />
+    <Compile Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Internals\SemanticConventions.cs" Link="Internals" />
+    <Compile Include="..\..\Azure.Monitor.OpenTelemetry.Exporter\src\Internals\StringExtensions.cs" Link="Internals" />
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Azure.Monitor.OpenTelemetry.LiveMetrics.csproj:29]: thank you! :)

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricConstants.cs [hunk 3]
```diff
@@ -17,5 +17,17 @@ internal static class LiveMetricConstants
         internal const string ExceptionsPerSecondInstrumentName = "ExceptionsPerSecondLiveMetric";
         internal const string MemoryCommittedBytesInstrumentName = "CommittedBytesLiveMetric";
         internal const string ProcessorTimeInstrumentName = "ProcessorTimeBytesLiveMetric";
+
+        internal const string RequestDurationMetricIdValue = @"\ApplicationInsights\Request Duration";
+        internal const string RequestsMetricIdValue = @"\ApplicationInsights\Requests/Sec";
+        internal const string RequestsSucceededPerSecondMetricIdValue = @"\ApplicationInsights\Requests Succeeded/Sec";
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricConstants.cs:23]: very nit: some of the conts have "PerSecond" in the name and others don't.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs [hunk 4]
```diff
@@ -48,5 +109,37 @@ protected override void Dispose(bool disposing)
...
+                                    Environment.GetEnvironmentVariable(WebSiteIsolationEnvironmentVariable) != WebSiteIsolationHyperV;
+                }
+                catch (Exception ex)
+                {
+                    LiveMetricsExporterEventSource.Log.AccessingEnvironmentVariableFailedWarning(WebSiteEnvironmentVariable, ex);
+                    return false;
+                }
+            }
+
+            return (bool)s_isAzureWebApp;
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs:137]: Can we avoid boxing here?
```suggestion
            return s_isAzureWebApp.Value;
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs [hunk 5]
```diff
@@ -48,5 +109,37 @@ protected override void Dispose(bool disposing)
...
+            if (!s_isAzureWebApp.HasValue)
+            {
+                try
+                {
+                    // Presence of "WEBSITE_SITE_NAME" indicate web apps.
+                    // "WEBSITE_ISOLATION"!="hyperv" indicate premium containers. In this case, perf counters
+                    // can be read using regular mechanism and hence this method retuns false for
+                    // premium containers.
+                    s_isAzureWebApp = !string.IsNullOrEmpty(Environment.GetEnvironmentVariable(WebSiteEnvironmentVariable)) &&
+                                    Environment.GetEnvironmentVariable(WebSiteIsolationEnvironmentVariable) != WebSiteIsolationHyperV;
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs:128]: TODO: switch to platform. Not necessary for POC.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs [hunk 6]
```diff
@@ -24,14 +33,66 @@ public LiveMetricsExporter(DoubleBuffer doubleBuffer, LiveMetricsExporterOptions
...
 
+            foreach (var metric in batch)
+            {
+                foreach (ref readonly var metricPoint in metric.GetMetricPoints())
+                {
+                    LiveMetricsExtractionProcessor.s_liveMetricNameMapping.TryGetValue(metric.Name, out var metricName);
+                    var liveMetricPoint = new Models.MetricPoint
+                    {
+                        Name = metricName
+                    };
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs:70]: Is the TryGet necessary here?
I don't see any checks if the TryGet was successful.


One optimization, I think the TryGet could be moved out of the second foreach.

**Review by TimothyMothra — APPROVED** [review]
LGTM. Left minor comments

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Internals/GuidExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Azure.Monitor.OpenTelemetry.LiveMetrics.Internals
+{
+    internal static class GuidExtensions
+    {
+        /// <summary>
+        /// Overload for Guid.ToString().
+        /// </summary>
+        /// <remarks>
+        /// This method encapsulates the language switch for NetStandard and NetFramework and resolves the error "The behavior of guid.ToStrinc() could vary based on the current user's locale settings".
```
Comment (vishweshbankwar) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Internals/GuidExtensions.cs:15]:
```suggestion
        /// This method encapsulates the language switch for NetStandard and NetFramework and resolves the error "The behavior of guid.ToString() could vary based on the current user's locale settings".
```

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricConstants.cs (hunk 3 above)
Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricConstants.cs:23]: RequestsMetricIdValue is the count of metrics. As the naming becomes complex avoided per second here. No strong opinion we could add to this one.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs [hunk 8]
```diff
@@ -24,14 +33,70 @@ public LiveMetricsExporter(DoubleBuffer doubleBuffer, LiveMetricsExporterOptions
...
+                {
+                    continue;
+                }
+
+                var liveMetricPoint = new Models.MetricPoint
+                {
+                    Name = metricName
+                };
+
+                foreach (ref readonly var metricPoint in metric.GetMetricPoints())
```
Comment (TimothyMothra) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs:74]: Just wanted to make one callout.
If this saves on allocations, then excellent :)

Reusing the liveMetricPoint might cause problems if the Add at line 94 is by ref. Should test to confirm that we're adding the full collection of metricpoints, or just several copies of the final metricpoint.

Comment (rajkumar-rangaraj) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs:74]: It should not have any impact, here is a similar implementation - https://github.com/Azure/azure-sdk-for-net/blob/c284046de29a1561373843a0f4f33600c474348a/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/MetricHelper.cs#L23

I agree, all of our implementation should be supported with test. Once we could get PoC working, we should add tests  on priority before adding new implementations.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs [hunk 9]
```diff
@@ -24,14 +33,70 @@ public LiveMetricsExporter(DoubleBuffer doubleBuffer, LiveMetricsExporterOptions
...
+            };
 
             foreach (var item in filledBuffer.ReadAllAndClear())
             {
                 monitoringDataPoint.Documents.Add(item);
             }
 
+            foreach (var metric in batch)
+            {
+                if (!LiveMetricsExtractionProcessor.s_liveMetricNameMapping.TryGetValue(metric.Name, out var metricName))
```
Comment (vishweshbankwar) [inline sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs:64]: Could you please add a comment why we are doing this here. May not be obvious for everyone.

File: sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs (hunk 8 above)
Comment (vishweshbankwar) [reply sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/LiveMetricsExporter.cs:74]: I think this won't work. We are overriding the same metricPoint. Am I missing something?

**Pull request merged**